cdn/signedurls/signurl.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2018 Google Inc. All rights reserved.
+// Use of this source code is governed by the Apache 2.0
+// license that can be found in the LICENSE file.
+
+// Command signedurls contains helper function to sign URLs.
+package main
+
+import (
+	"crypto/hmac"
+	"crypto/sha1"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"time"
+)
+
+// [START example]
+
+// SignURL creates a signed URL for an endpoint on Cloud CDN. sourceURL must
+// start with "https://" and should not have the Expires, KeyName, or Signature
+// query parameters. key should be in raw form (not base64url-encoded).
+func SignURL(sourceURL, keyName string, key []byte, expiration time.Time) string {
+	u := sourceURL
+	sep := "?"
+	if strings.Contains(u, "?") {
+		sep = "&"
+	}
+	u += sep
+	u += fmt.Sprintf("Expires=%d", expiration.Unix())
+	u += fmt.Sprintf("&KeyName=%s", keyName)
+
+	sig := base64urlencode(hmacSHA1(u, key))
+	u += fmt.Sprintf("&Signature=%s", sig)
+	return u
+}
+
+func hmacSHA1(msg string, key []byte) []byte {
+	mac := hmac.New(sha1.New, key)
+	mac.Write([]byte(msg))
+	return mac.Sum(nil)
+}
```

Comment:
just inline this function

Reply:
I think this provides clear separation and makes the algorithm in signURL clearer by simply providing a way to do `f1(f2(val))`.

Reply:
On the other hand, it turns it into two functions, rather than just the one. If you're including just the `SignURL` function in docs, then one function is better.

If someone is copying and pasting, then just one function is better.

Reply:
My current plan is to include the entire file excl. the license header. There's a lot of encoding/decoding is going on, I think separation helps the user understand the algorithm better as well.